lib/prediction/localparserclient.ts

```diff
@@ -206,27 +212,31 @@ export default class LocalParserClient {
         }
 
         if (result === null) {
-            if (options.expect === 'Location') {
-                result = [{
-                    code: ['$answer', '(', 'new', 'Location', '(', '"', ...tokens, '"', ')', ')', ';'],
-                    score: 1
-                }];
-            } else {
-                if (contextCode)
-                    contextCode = this._applyPreHeuristics(contextCode);
+            if (contextCode)
+                contextCode = this._applyPreHeuristics(contextCode);
 
-                let candidates;
-                if (contextCode)
-                    candidates = await this._predictor.predict(contextCode.join(' '), tokens.join(' '), answer, NLU_TASK, options.example_id);
-                else
-                    candidates = await this._predictor.predict(tokens.join(' '), undefined, answer, SEMANTIC_PARSING_TASK, options.example_id);
-                result = candidates.map((c) => {
-                    return {
-                        code: c.answer.split(' '),
-                        score: c.score
-                    };
-               });
-            }
+            let candidates;
+            if (contextCode)
+                candidates = await this._predictor.predict(contextCode.join(' '), tokens.join(' '), answer, NLU_TASK, options.example_id);
+            else
+                candidates = await this._predictor.predict(tokens.join(' '), undefined, answer, SEMANTIC_PARSING_TASK, options.example_id);
+            assert(candidates.length > 0);
+
+            result = candidates.map((c) => {
+                // convert is_correct and is_probably_correct scores into
+                // a single scale such that >0.5 is correct and >0.25 is
+                // probably correct
+                const score = (c.score.is_correct ?? 1) > 0.5 ? (c.score.is_correct ?? 1) :
```

Comment:
L229-L239: I don't understand why you are doing calculations with these numbers.

Reply:
These are not probably scores. As I said above, they are not guaranteed to be in [0, 1].

Reply:
These are meant to be used as follows:

```
if (is_junk> threshold) {
  ignore;
}
elif (is_ood > threshold){
  send to different backends;
elif (is_probably_correct < threshold) {
 parsing error;
}
elif (is_correct < threshold) {
 probably correct, but confirm with user to be sure;
}
else {
  correct;
}
```

Reply:
Well, the API that we provide to clients has one confidence number per candidate parse, in the range [0, 1], and a set of intent confidence scores, each in the range [0,1] and summing to 1.
So please adjust them here or in genienlp.